File: cpp11/Ice/async/HelloI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#include <Ice/Ice.h>
#include <HelloI.h>

using namespace std;

HelloI::HelloI(const shared_ptr<WorkQueue>& workQueue) :
    _workQueue(workQueue)
{
}

void
HelloI::sayHelloAsync(int delay,
                      function<void()> response,
                      function<void(exception_ptr)> error,
                      const Ice::Current&)
{
    if(delay == 0)
    {
        cout << "Hello World!" << endl;
        response();
    }
    else
    {
        _workQueue->add(delay, move(response), move(error));
    }
}

void
HelloI::shutdown(const Ice::Current& curr)
{
    cout << "Shutting down..." << endl;

    _workQueue->destroy();
    curr.adapter->getCommunicator()->shutdown();
}

File: java-compat/Ice/async/HelloI.java
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// **********************************************************************

import Demo.*;

public class HelloI extends _HelloDisp
{
    public 
    HelloI(WorkQueue workQueue)
    {
        _workQueue = workQueue;
    }

    @Override
    public void
    sayHello_async(AMD_Hello_sayHello cb, int delay, Ice.Current current)
    {
        if(delay == 0)
        {
            System.out.println("Hello World!");
            cb.ice_response();
        }
        else
        {
            _workQueue.add(cb, delay);
        }
    }

    @Override
    public void
    shutdown(Ice.Current current)
    {
        System.out.println("Shutting down...");

        _workQueue._destroy();
        current.adapter.getCommunicator().shutdown();
    }

    private WorkQueue _workQueue;
}
